0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+encoding its data either in UTF-16 or in Latin-1, where the latter is an optimization for space
+when all characters in the string are supported by this encoding. 
+
+Java does not included a dedicated type for Unicode scalar values
+and instead uses the 32-bit wide `int` primitive type.
+
+## Go:
+
+In Go a string is a read-only slice of bytes, which generally contains UTF-8 encoded bytes
+but may contain arbitrary bytes. The `rune` type is an alias for `int32`.
+
+## JavaScript:
+
+The JavaScript `String` type represents an immutable sequence of UTF-16 code units as integer values.
+There is no distinct type for representing individual characters.
+
+## C++:
+
+The `char` type represents bytes, and the `std::string` class from the standard library
+operates on bytes as character, and generally does not produce correct results if used
+together with any encoding other than single-byte encodings such as ASCII.
+C++11 added two new character types, `char16_t` and `char32_t`, 
+and two new corresponding `std::u16string` and `std::u32string` collection classes.
+It also provides three new kinds of string literals,
+`u8"..."`, `u"..."`, and `U"..."`,
+which produce binary values encoded with UTF-8, UTF-16, and UTF-32 respectively.
+
+## Python:
+
+In Python 2, the `str` type was used for both binary and text data.
+A separate `unicode` type was introduced to store textual data,
+and a separate mode for unicode string literals (`u"..."`).
+Python 3 changed `str` to be this `unicode` datatype instead.
+Like JavaScript, Python has no dedicated character type,
+and just uses the unbounded `int` type for Unicode scalar values.
+
+## Rust:
+
+The `char` type directly represents any Unicode scalar value, and rejects out-of-range values.
+The "string slice" type `str` is the most primitive string type,
+which is guaranteed to refers to valid UTF-8 data.
+It primarily acts as a sequence of bytes, but directly provides many methods that work with `char` values:
+`chars()` provides an iterator of `char` values,
+`get(<slice>)` returns a subslice but produces `None` if the bounds do not align with UTF-8 sequence boundaries,
+etc.
+
+## Swift:
+
+Swift defines `String` as a collection of characters,
+and unusually defines `Character` as an [*extended grapheme cluster*](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries),
+which is the closest concept to an atomic characeter as perceived visually by users.
+Unicode scalar values are represented with a distinct `Unicode.Scalar` structure instead.
+`String` provides `utf8`, `utf16`, and `unicodeScalars` views
+that present the contents as sequences of UTF-8, UTF-16, and UTF-32 code units respectively.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Is there anything more we can do to make migration easier and safer for users?
+Chance are very good that all Dafny code in existence to date either will not change behavior
```

Comment:
This seems overly optimistic to me WRT string literals, but I'm not sure how we plan to handle them.  At the moment they are sequences of characters mixed with UTF16 code units (possibly ill-formed), right? It would be good to specify what happens to them in `/unicodeChar:1`

Reply:
I think this is moot now but let me know otherwise.

Reply:
There will be code that we won't be able to port (code using string literals with unpaired surrogates), but that's fine

Reply:
Yup, the worst case scenario will be code using `string` to carry such data, which the conversion utility (or verification with `/unicodeChar:1`) will reject, and the code will have to be rewritten to use `seq<uint16>` instead. I'm definitely okay with that, especially as I think it's much more likely such code will be unaware of the issue with unpaired surrogates rather than intentionally using the current definition of `string` to allow them.